crates/oxc_codegen/src/code_buffer.rs

```diff
@@ -0,0 +1,421 @@
+        // SAFETY: `buf` is a valid UTF-8 string because of invariants upheld by CodeBuffer
+        unsafe { std::str::from_utf8_unchecked(&self.buf) }.chars().nth_back(n)
+    }
+
+    /// Push a single ASCII character into the buffer
+    ///
+    /// # Panics
+    /// If `ch` is not a valid UTF-8 code point in the ASCII range (`0 - 0x7F`).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use oxc_codegen::CodeBuffer;
+    /// let mut code = CodeBuffer::new();
+    /// code.print_ascii_char('f');
+    /// code.print_ascii_char('o');
+    /// code.print_ascii_char('o');
+    ///
+    /// let source = code.take_source_text();
+    /// assert_eq!(source, "foo");
+    /// ```
+    #[inline]
+    pub fn print_ascii_char(&mut self, ch: u8) {
+        // NOTE: since this method is inlined, this assertion should get
+        // optimized away by the compiler when the value of `ch` is known, e.g.
+        // when printing a constant.
+        assert!(ch.is_ascii(), "character {ch} is not ASCII");
+        self.buf.push(ch);
+    }
+
+    /// Print a byte without checking that this buffer still represents a valid
+    /// UTF-8 string.
+    ///
+    /// If you are looking to print a byte you know is valid ASCII, prefer
+    /// [`print_ascii_char`]. If you are not certain, you may use [`print_char`]
+    /// as a safe alternative.
+    ///
+    /// # Safety
+    /// The caller must ensure that, after 1 or more sequential calls, this
+    /// buffer represents a valid UTF-8 string.
+    ///
+    /// It is safe for a single call to temporarily result in invalid UTF-8, as
+    /// long as UTF-8 integrity is restored before calls to any other `print`
+    /// method or [`take_source_text`]. This lets you, for example, print an
+    /// 8-byte code point using 4 separate calls to this method.
+    ///
+    /// If you find yourself in such a scenario, consider using
+    /// [`print_unchecked`] instead.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use oxc_codegen::CodeBuffer;
+    /// let mut code = CodeBuffer::new();
+    /// // Safe: 'a' is a valid ASCII character. Its UTF-8 representation only
+    /// // requires a single byte.
+    /// unsafe { code.print_byte(b'a') };
+    ///
+    /// let not_ascii = '⚓';
+    /// let as_bytes = not_ascii.to_string().into_bytes();
+    /// // Safe: after this loop completes, `code` returns to a valid state.
+    /// for byte in as_bytes {
+    ///     unsafe { code.print_byte(byte) };
+    /// }
+    ///
+    /// // NOT SAFE: `ch` exceeds the ASCII segment range. `code` is no longer
+    /// valid UTF-8
+    /// unsafe { code.print_byte(0xFF) };
+    /// ```
+    ///
+    /// [`print_ascii_char`]: CodeBuffer::print_ascii_char
+    /// [`print_char`]: CodeBuffer::print_char
+    /// [`take_source_text`]: CodeBuffer::take_source_text
+    /// [`print_unchecked`]: CodeBuffer::print_unchecked
+    #[inline]
+    pub unsafe fn print_byte(&mut self, ch: u8) {
```

Comment:
I think this should be called `print_byte_unchecked` to reflect that it's an unsafe method.